### PR TITLE
Fix admin template setters

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

This PR will revert some deprecations to allow override `Admin` templates in old way.
Also override templates by `TemplateRegistry::setTemplates()` should not replace but override it. People can call it in own code and break application working.

I am targeting this branch, because this change respect BC.

## Changelog

```markdown
### Added
- Added `Sonata\AdminBundle\Templating\MutableTemplateRegistryAwareInterface::setTemplate()`
- Added `Sonata\AdminBundle\Templating\MutableTemplateRegistryAwareInterface::setTemplates()`
### Removed
- Removed deprecations for `Sonata\AdminBundle\Admin\AbstractAdmin::setTemplate()`
- Removed deprecations for `Sonata\AdminBundle\Admin\AbstractAdmin::setTemplates()`
```

### CODE REVIEW

**Comment by wbloszyk** [comment]
Mybe we can create `MutableTemplateRegistry` in `__constructor()` and set templates by `setTemplate()` and `setTemplates()` instead creating new services? IMHO `setTemplateRegistry()` is unnesessery.

https://github.com/sonata-project/SonataAdminBundle/blob/7e0bdd2c3e7c63f3c6495442149d94ca2dd143f8/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L403-L415

WDYT? @sonata-project/contributors

File: src/Templating/MutableTemplateRegistry.php [hunk 1]
```diff
@@ -22,7 +22,7 @@ final class MutableTemplateRegistry extends AbstractTemplateRegistry implements
     // public function setTemplates(array $templates): void
     public function setTemplates(array $templates)
     {
-        $this->templates = $templates;
+        $this->templates = array_merge($this->templates, $templates);
```
Comment (VincentLanglet) [inline src/Templating/MutableTemplateRegistry.php:25]: Is this change needed ? Seems like a BC-break and change a lot this method.

According to https://symfony.com/doc/current/bundles/SonataAdminBundle/reference/templates.html
```
A setTemplates(array $templates) (notice the plural) method also exists, that allows you to set multiple templates at once. Notice that, if used outside of the service definition context, setTemplates(array $templates) will replace the whole template list for that Admin class, meaning you have to explicitly pass the full template list in the $templates argument.
```

So it seems expected to override the whole template list to me.

There is already 
```
if ('setTemplates' === $method) {
                $definedTemplates = array_merge($definedTemplates, $args[0]);

                continue;
            }
```
In the AddDependencyCallsCompilerPass

Comment (wbloszyk) [reply src/Templating/MutableTemplateRegistry.php:25]: We can change methods for `overrideTemplate()` and `overrideTemplates()` but it is bad idea.

When we call method `setTemplates()` in service config, we will use `array_merge`.
When someone call `AbstractAdmin::setTemplates()` in code then templates will be replace.

The same method work in two diffrence way.

We should fix it. If we cant do it now, it should be mark as `NEXT_MAJOR`.

Comment (VincentLanglet) [reply src/Templating/MutableTemplateRegistry.php:25]: I don't think someone will use it to override every template at once.
So I think the array_merge is not a bad idea.

But since it's a BC-break, I think the `NEXT_MAJOR: use array merge and update the doc https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/reference/templates.rst` is better indeed.

Comment (wbloszyk) [reply src/Templating/MutableTemplateRegistry.php:26]: Done.

File: src/Templating/MutableTemplateRegistry.php [hunk 2]
```diff
@@ -22,7 +22,9 @@ final class MutableTemplateRegistry extends AbstractTemplateRegistry implements
     // public function setTemplates(array $templates): void
     public function setTemplates(array $templates)
     {
-        $this->templates = $templates;
+        //NEXT_MAJOR: use array merge and update the doc https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/reference/templates.rst
+        $this->templates = array_merge($this->templates, $templates);
```
Comment (VincentLanglet) [inline src/Templating/MutableTemplateRegistry.php:26]: U let the line

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: src/Templating/MutableTemplateRegistry.php (hunk 2 above)
Comment (wbloszyk) [reply src/Templating/MutableTemplateRegistry.php:26]: What i done. :laughing:

**Review by VincentLanglet — APPROVED** [review]

File: tests/Templating/MutableTemplateRegistryTest.php [hunk 3]
```diff
@@ -44,7 +44,9 @@ public function testGetTemplates(): void
         ];
 
         $this->templateRegistry->setTemplates($templates);
+        // NEXT_MAJOR: remove this first assert and uncomment second
         $this->assertSame($templates, $this->templateRegistry->getTemplates());
+        //$this->assertSame(array_merge(['list' => '@FooAdmin/CRUD/list.html.twig'], $templates), $this->templateRegistry->getTemplates());
```
Comment (phansys) [inline tests/Templating/MutableTemplateRegistryTest.php:49]:
```suggestion
        //$this->assertSame($templates + ['list' => '@FooAdmin/CRUD/list.html.twig'], $this->templateRegistry->getTemplates());
```

File: src/Templating/MutableTemplateRegistry.php [hunk 4]
```diff
@@ -22,7 +22,9 @@ final class MutableTemplateRegistry extends AbstractTemplateRegistry implements
     // public function setTemplates(array $templates): void
     public function setTemplates(array $templates)
     {
+        //NEXT_MAJOR: use array merge and update the doc https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/reference/templates.rst
         $this->templates = $templates;
+        //$this->templates = array_merge($this->templates, $templates);
```
Comment (phansys) [inline src/Templating/MutableTemplateRegistry.php:27]:
```suggestion
        //$this->templates = $templates + $this->templates;
```

**Review by phansys — CHANGES_REQUESTED** [review]

**Comment by wbloszyk** [comment]
@sonata-project/contributors @phansys Can we merge it?

File: src/Templating/MutableTemplateRegistry.php [hunk 5]
```diff
@@ -22,7 +22,9 @@ final class MutableTemplateRegistry extends AbstractTemplateRegistry implements
     // public function setTemplates(array $templates): void
     public function setTemplates(array $templates)
     {
+        //NEXT_MAJOR: use array merge and update the doc https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/reference/templates.rst
```
Comment (phansys) [inline src/Templating/MutableTemplateRegistry.php:25]:
```suggestion
        //NEXT_MAJOR: merge arrays and update the doc https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/reference/templates.rst
```

**Review by phansys — DISMISSED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @wbloszyk

**Comment by VincentLanglet** [comment]
> Mybe we can create `MutableTemplateRegistry` in `AbstractAdmin::__constructor()` and set templates by `setTemplate()` and `setTemplates()` instead creating new services? IMHO `setTemplateRegistry()` is unnesessery and add some confusing problems like how to set own templates, what is impossible for now.
> 
> In master (4.0) we dont use template registry services but take it directly from `Admin`.
> 
> https://github.com/sonata-project/SonataAdminBundle/blob/7e0bdd2c3e7c63f3c6495442149d94ca2dd143f8/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L403-L415
> 
> WDYT? @sonata-project/contributors

Not sure to understand everything and if it's related, but isn't it weird now to have a `setTemplate` method in both `MutableTemplateRegistryAwareInterface` and `MutableTemplateRegistryInterface` ?
I feel like the `MutableTemplateRegistryInterface` is useless now.

What would be the next @wbloszyk ?

**Comment by wbloszyk** [comment]
Next step will be add `MutableTemplateRegistryAwareInterface::setTemplatesByTemplateRegistry($registry, $override = true)`. Override false will be use to set global templates by compiler pass which is call after admin service configuration.

In 3.x template registry is created by compiler pass and can not be override(it is possibile but hard to do).

In 4.x template registry will be create by admin constructor. Override template registry will be not allowed but override templates by registry will be easy. 


In 5.x I will try to present new templates registry system
